modules/capacity-reservation/main.tf

```diff
@@ -0,0 +1,67 @@
+locals {
+  capacity_reservations = merge(flatten([
+    for ng_name, ng in var.instance_capacity : [
+      for az_id in ng.availability_zone_ids : {
+        for type in ng.instance_types : "${type}-${az_id}" => {
+          capacity      = ng.capacity
+          instance_type = type
+          az            = data.aws_availability_zone.zone_ids[az_id].name
+        }
+      }
+    ]
+  ])...)
+  zone_ids = toset(flatten([for ng in var.instance_capacity :
+    ng.availability_zone_ids
+  ]))
+
+  instance_types = toset([for cr in local.capacity_reservations : cr.instance_type])
+}
+
+data "aws_availability_zone" "zone_ids" {
+  for_each = local.zone_ids
+  zone_id  = each.key
+}
+
+
+data "aws_ec2_instance_type_offerings" "this" {
+  for_each = local.instance_types
+
+  filter {
+    name   = "instance-type"
+    values = [each.key]
+  }
+
+  location_type = "availability-zone"
+}
+
+
+resource "aws_ec2_capacity_reservation" "this" {
```

Comment:
I didn't log this as a change request, but I'm not 100% behind using "this" as an identifier for resources. I understand some things like aws regions, etc., in datasources do that, but it still feels weird and inspecific.

Reply:
each resource will be identified by the key in the `for_each`